Fix costmodel creation in Alonzo Genesis

teodanciu opened the pull request.

teodanciu committed:
Use hardcoded parameter names to sort the costmodel parameter values

instead of the plutus `ParamName` enum, since the latter returns
slightly different names for 6 parameters:

"blake2b_256-cpu-arguments-intercept",
"blake2b_256-cpu-arguments-slope",
"blake2b_256-memory-arguments",
"verifyEd25519Signature-cpu-arguments-intercept",
"verifyEd25519Signature-cpu-arguments-slope",
"verifyEd25519Signature-memory-arguments"

Instead of the names expected by the alonzo genesis file:
"blake2b-cpu-arguments-intercept",
"blake2b-cpu-arguments-slope",
"blake2b-memory-arguments",
"verifySignature-cpu-arguments-intercept",
"verifySignature-cpu-arguments-slope",
"verifySignature-memory-arguments"

Also, replace the `FromJSON` instance with a function, to make it clear
that it's a legacy implementation.

respectively.

teodanciu committed:
Add a golden test to verify the deserialization of Alonzo Genesis file

and the sorting of cost model parameters

JaredCorduan commented on eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal:
```diff
@@ -18,6 +18,7 @@ data-files:
   cddl-files/real/crypto.cddl
   cddl-files/mock/extras.cddl
   golden/*.cbor
+  golden/mainnet-alonzo-genesis.json
```
:heart:

JaredCorduan reviewed (approved):
This looks great to me, thank you @teodanciu !

@Jimbo4350 - you should be aware that the CLI is going to need to use `legacyParseCostModels` for parsing the original alonozo genesis file.

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs:
```diff
@@ -216,38 +211,30 @@ instance ToJSONKey Language where
+  do
+    plutusV1 <- o .:? "PlutusV1"
+    cms <- traverse (validateCostModel PlutusV1 . cmParamValues) plutusV1
+    pure . CostModels . Map.fromList $ maybeToList cms
```
@teodanciu - I talked with @Jimbo4350 and learned how node is using the Alonzo genesis for testing. They want the ability to load up a bunch of cost models (V2, etc) in the alonzo era. Let's add parsing for the other language versions, but have them all just parse a list of integers (instead of the map).

teodanciu replied:
Ah, sorry, just saw your comment.  
I just added a commit that parses V1 and V2 in the ToJSON instance of CostModels, that is not used by the code (only by an example NewGenesis data type that I added in tests). 

Do you know if they need the parsing of AlonzoGenesis or that of CostModels (that is contained within the AlonzoGenesis)? 
1)  If they need only the CostModels deserialization, then this new commit that I added should be fine - they only need to call FromJSON and will get CostModels for both versions.
2) However, if they want to deserialize AlonzoGenesis and expect CostModels for both V1 and V2, then I can make a simple change and not make the differentiation between legacy and non-legacy: let the FromJSON instance parse both V1 and V2 and construct AlonzoGenesis with both. 

What do you think between 1) and 2)

Sorry for the long comment - i hope it makes some sense.

JaredCorduan replied:
I think we need 2 for testing purposes.

teodanciu replied:
Ok!

teodanciu committed:
Support parsing of PlutusV2 costmodel params in `AlonzoGenesis`

which is needed for the future, non-legacy parsing of genesis and also
for current testing.

teodanciu commented:
Hello @Jimbo4350 - would you mind taking a look at this before we're merging it? Just to make sure it will  work with your plans for testing.

Jimbo4350 reviewed (approved):
LGTM! Thanks :+1:
